File: src/main/java/com/somefrills/features/misc/glowmob/GlowMob.java
package com.somefrills.features.misc.glowmob;

import com.somefrills.config.FrillsConfig;
import com.somefrills.config.misc.MobGlowConfig.GlowMobRule;
import com.somefrills.events.GameStopEvent;
import com.somefrills.events.TickEventPost;
import com.somefrills.features.core.Feature;
import com.somefrills.misc.RenderColor;
import com.somefrills.misc.Utils;
import meteordevelopment.orbit.EventHandler;
import meteordevelopment.orbit.EventPriority;
import net.minecraft.entity.LivingEntity;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class GlowMob extends Feature {

    private final List<GlowMobRule> rules;
    private List<LivingEntity> entityList = List.of();

    public GlowMob() {
        super(FrillsConfig.instance.misc.glowMob.enabled);
        rules = FrillsConfig.instance.misc.glowMob.rules;
    }

    /* ---------------- ENTITY CACHE ---------------- */

    private List<LivingEntity> getEntities() {
        if (entityList == null) updateEntities();
        return entityList;
    }

    private void updateEntities() {
        entityList = Utils.getEntities().stream()
                .filter(Utils::isMob)
                .map(LivingEntity.class::cast)
                .toList();
    }

    /* ---------------- CORE MATCHING ---------------- */

    private RenderColor findGlowMatch(LivingEntity living) {
        for (GlowMobRule rule : rules) {
            if (rule == null || !rule.enabled()) continue;
            if (rule.matches(living)) return rule.color();
        }
        return null;
    }

    private void applyHighlight(LivingEntity living) {
        RenderColor color = findGlowMatch(living);
        if (color != null) {
            Utils.setGlowing(living, true, color);
        }
    }

    private void forEachMatching(GlowMobRule rule, Consumer<LivingEntity> action) {
        if (rule == null) return;

        for (LivingEntity living : getEntities()) {
            if (rule.matches(living)) {
                action.accept(living);
            }
        }
    }

    private void clearGlow(LivingEntity living) {
        Utils.setGlowing(living, false, RenderColor.white);
    }

    private void applyGlow(LivingEntity living, GlowMobRule rule) {
        Utils.setGlowing(living, true, rule.color());
    }

    /* ---------------- LIFECYCLE ---------------- */

    @Override
    protected void onDeactivate() {
        updateEntities();
        for (LivingEntity living : getEntities()) {
            for (GlowMobRule rule : rules) {
                if (rule.matches(living)) {
                    clearGlow(living);
                    break;
                }
            }
        }
    }

    @EventHandler(priority = EventPriority.LOW)
    private void onWorldTick(TickEventPost event) {
        updateEntities();
        getEntities().forEach(this::applyHighlight);
    }

    @EventHandler
    private void onGameClose(GameStopEvent event) {
        if (FrillsConfig.instance.misc.glowMob.saveRules) return;
        clearRules();
    }

    /* ---------------- RULE MANAGEMENT ---------------- */

    public int addRule(GlowMobRule rule) {
        if (!isActive() || rule == null) return -1;

        try {
            rules.add(rule);
            return rules.size();
        } catch (Exception e) {
            Utils.infoFormat("Failed to add glow rule: {}", e.getMessage());
            return -1;
        }
    }

    public int addRule(MatchInfo matcher, RenderColor color) {
        return addRule(new GlowMobRule(matcher, color));
    }

    public boolean removeRule(GlowMobRule rule) {
        return removeRule(rules.indexOf(rule) + 1);
    }

    public boolean removeRule(int id) {
        if (!isActive()) return false;

        GlowMobRule removed = rules.remove(id - 1);
        if (removed == null) return false;

        updateEntities();
        forEachMatching(removed, this::clearGlow);

        return true;
    }

    public void replaceRule(GlowMobRule original, GlowMobRule replacement) {
        if (!isActive()) return;

        int idx = rules.indexOf(original);
        if (idx == -1) {
            Utils.info("Original rule not found, cannot replace");
            return;
        }

        rules.set(idx, replacement);
        updateEntities();

        for (LivingEntity living : getEntities()) {
            boolean oldMatch = original.matches(living);
            boolean newMatch = replacement.matches(living);

            if (oldMatch && !newMatch) {
                clearGlow(living);
            } else if (!oldMatch && newMatch) {
                applyGlow(living, replacement);
            }
        }
    }

    public void clearRules() {
        if (!isActive()) return;

        updateEntities();

        for (GlowMobRule rule : rules) {
            forEachMatching(rule, this::clearGlow);
        }

        rules.clear();
    }

    public void toggleRule(int parsedId) {
        if (!isActive()) return;

        GlowMobRule rule = rules.get(parsedId - 1);
        if (rule == null) return;

        updateEntities();

        if (rule.enabled()) {
            forEachMatching(rule, this::clearGlow);
        } else {
            forEachMatching(rule, living -> applyGlow(living, rule));
        }

        rule.toggle();
    }

    /* ---------------- QUERY API ---------------- */

    public List<GlowMobRule> getRules() {
        return rules;
    }

    public static class MatchedEntityEntry {
        public GlowMobRule rule;
        public List<LivingEntity> entities;

        public MatchedEntityEntry(GlowMobRule rule, List<LivingEntity> entities) {
            this.rule = rule;
            this.entities = entities;
        }
    }

    public List<MatchedEntityEntry> getGlowingMobs() {
        return getGlowingMobs(rules);
    }

    public List<MatchedEntityEntry> getGlowingMobs(List<GlowMobRule> rules) {
        updateEntities();

        ArrayList<MatchedEntityEntry> result = new ArrayList<>();

        for (GlowMobRule rule : rules) {
            if (!rule.enabled()) continue;

            List<LivingEntity> matchedEntities = getEntities().stream()
                    .filter(rule::matches)
                    .toList();

            result.add(new MatchedEntityEntry(rule, matchedEntities));
        }

        return result;
    }
}

File: src/main/kotlin/com/somefrills/features/misc/glowmob/GlowMob.kt
package com.somefrills.features.misc.glowmob

import com.somefrills.config.FrillsMod

import com.somefrills.config.misc.GlowMobConfig.GlowMobRule
import com.somefrills.events.GameStopEvent
import com.somefrills.events.TickEventPost
import com.somefrills.features.core.Feature
import com.somefrills.modules.FrillsFeature
import com.somefrills.misc.RenderColor
import com.somefrills.utils.ChatUtils
import com.somefrills.utils.EntityUtils
import com.somefrills.utils.isMob
import com.somefrills.utils.setGlowing
import com.somefrills.events.core.EventHandle
import com.somefrills.events.core.EventPriority
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.LivingEntity
import java.util.function.Consumer

@FrillsFeature
object GlowMob : Feature(FrillsMod.config.misc.glowMob.enabled) {
    private val config get() = FrillsMod.config.misc.glowMob

    /* ---------------- QUERY API ---------------- */
    @JvmField
    val rules: MutableList<GlowMobRule> = config.rules
    private var entityList: MutableList<LivingEntity>? = null
    private val entities: MutableList<LivingEntity>
        get() {
            if (entityList == null) updateEntities()
            return entityList!!
        }

    private fun updateEntities() {
        entityList = EntityUtils.getStreamEntities(LivingEntity::class.java)
            .filter { entity: Entity -> entity.isMob() }
            .toMutableList()
    }

    /* ---------------- CORE MATCHING ---------------- */
    private fun findGlowMatch(living: LivingEntity): RenderColor? {
        for (rule in rules) {
            if (!rule.enabled()) continue
            if (rule.matches(living)) return rule.color()
        }
        return null
    }

    private fun applyHighlight(living: LivingEntity) {
        val color = findGlowMatch(living)
        if (color != null) {
            living.setGlowing(true, color)
        }
    }

    private fun forEachMatching(rule: GlowMobRule, action: Consumer<LivingEntity>) {
        for (living in this.entities) {
            if (rule.matches(living)) {
                action.accept(living)
            }
        }
    }

    private fun clearGlow(living: LivingEntity) {
        living.setGlowing(false, RenderColor.white)
    }

    private fun applyGlow(living: LivingEntity, rule: GlowMobRule) {
        living.setGlowing(true, rule.color())
    }

    /* ---------------- LIFECYCLE ---------------- */
    override fun onDeactivate() {
        updateEntities()
        for (living in this.entities) {
            for (rule in rules) {
                if (rule.matches(living)) {
                    clearGlow(living)
                    break
                }
            }
        }
    }

    @EventHandle(priority = EventPriority.LOW)
    private fun onWorldTick(event: TickEventPost) {
        updateEntities()
        this.entities.forEach(Consumer { living: LivingEntity -> this.applyHighlight(living) })
    }

    @EventHandle
    private fun onGameClose(event: GameStopEvent) {
        if (config.saveRules) return
        clearRules()
    }

    /* ---------------- RULE MANAGEMENT ---------------- */
    fun addRule(rule: GlowMobRule): Int {
        if (!isActive()) return -1

        try {
            rules.add(rule)
            return rules.size
        } catch (e: Exception) {
            ChatUtils.infoFormat("Failed to add glow rule: {}", e.message)
            return -1
        }
    }

    fun addRule(matcher: MatchInfo, color: RenderColor): Int {
        return addRule(GlowMobRule(matcher, color))
    }

    fun removeRule(rule: GlowMobRule): Boolean {
        return removeRule(rules.indexOf(rule) + 1)
    }

    fun removeRule(id: Int): Boolean {
        if (!isActive()) return false

        val removed = rules.removeAt(id - 1)

        updateEntities()
        forEachMatching(removed) { living: LivingEntity -> this.clearGlow(living) }

        return true
    }

    fun replaceRule(original: GlowMobRule, replacement: GlowMobRule) {
        if (!isActive()) return

        val idx = rules.indexOf(original)
        if (idx == -1) {
            ChatUtils.info("Original rule not found, cannot replace")
            return
        }

        rules[idx] = replacement
        updateEntities()

        for (living in this.entities) {
            val oldMatch = original.matches(living)
            val newMatch = replacement.matches(living)

            if (oldMatch && !newMatch) {
                clearGlow(living)
            } else if (!oldMatch && newMatch) {
                applyGlow(living, replacement)
            }
        }
    }

    fun clearRules() {
        if (!isActive()) return

        updateEntities()

        for (rule in rules) {
            forEachMatching(rule) { living: LivingEntity -> this.clearGlow(living) }
        }

        rules.clear()
    }

    fun toggleRule(parsedId: Int) {
        if (!isActive()) return

        val rule = rules[parsedId - 1]

        updateEntities()

        if (rule.enabled()) {
            forEachMatching(rule) { living: LivingEntity -> this.clearGlow(living) }
        } else {
            forEachMatching(rule) { living: LivingEntity -> applyGlow(living, rule) }
        }

        rule.toggle()
    }

    class MatchedEntityEntry(@JvmField var rule: GlowMobRule, @JvmField var entities: MutableList<LivingEntity>)


    @JvmStatic
    fun getGlowingMobs(): MutableList<MatchedEntityEntry> {
        return getGlowingMobs(rules)
    }

    @JvmStatic
    fun getGlowingMobs(rules: List<GlowMobRule>): MutableList<MatchedEntityEntry> {
        updateEntities()

        val result = ArrayList<MatchedEntityEntry>()

        for (rule in rules) {
            if (!rule.enabled()) continue

            val matchedEntities = this.entities.stream()
                .filter { entity: LivingEntity -> rule.matches(entity) }
                .toList()

            result.add(MatchedEntityEntry(rule, matchedEntities))
        }

        return result
    }
}
